### PR TITLE
Add DbRef Feature And Test

### PR DESCRIPTION
**Pull request by humhei**
@Zaid-Ajaj  Hello, May I contribute to this Repository  ?

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,5 @@
 bin
 obj
 packages
-paket-files
\ No newline at end of file
+paket-files
```
Comment (humhei) [inline .gitignore]: Update LiteDB Package  to 4.0

File: LiteDB.FSharp.Tests/LiteDB.FSharp.Tests.fsproj [hunk 2]
```diff
@@ -7,7 +7,8 @@
   <ItemGroup>
     <Compile Include="Tests.Types.fs" />
     <Compile Include="Tests.Bson.fs" />
-    <Compile Include="Tests.LiteDatabase.fs" />
+    <Compile Include="Tests.LiteDatabase.fs" />  
+    <Compile Include="Tests.DBRef.fs" />
```
Comment (humhei) [inline LiteDB.FSharp.Tests/LiteDB.FSharp.Tests.fsproj]: Add DBRef Test

File: LiteDB.FSharp.Tests/Tests.Types.fs [hunk 3]
```diff
@@ -29,4 +28,17 @@ type ComplexUnion<'t> =
     | Any of 't
     | Int of int
     | String of string 
-    | Generic of Maybe<'t>
```
Comment (humhei) [inline LiteDB.FSharp.Tests/Tests.Types.fs]: Add CLIMutable Type

File: LiteDB.FSharp/Bson.fs [hunk 4]
```diff
@@ -58,20 +58,57 @@ module Bson =
 
     /// Converts a BsonDocument to a typed entity given the document the type of the CLR entity.
```
Comment (humhei) [inline LiteDB.FSharp/Bson.fs]: Rewrite deserializeByType So Can Resolve NestedId Currently

File: LiteDB.FSharp/FSharpBsonMapper.fs [hunk 5]
```diff
@@ -1,9 +1,29 @@
...
+open System
+open System.Collections.Generic
+
 
 type FSharpBsonMapper() = 
     inherit BsonMapper()
+    let entityMappers=Dictionary<Type,EntityMapper>() 
     override self.ToObject(entityType: System.Type, entity: BsonDocument) = Bson.deserializeByType entity entityType 
     override self.ToObject<'t>(entity: BsonDocument) = Bson.deserialize<'t> entity
-    override self.ToDocument<'t>(entity: 't) = Bson.serialize entity
```
Comment (humhei) [inline LiteDB.FSharp/FSharpBsonMapper.fs]: Add withEntityMap Function to support DBref

File: LiteDB.FSharp.Tests/Tests.Types.fs [hunk 6]
```diff
@@ -29,4 +28,28 @@ type ComplexUnion<'t> =
...
\ No newline at end of file
+    | Generic of Maybe<'t>
```
Comment (humhei) [inline LiteDB.FSharp.Tests/Tests.Types.fs]: Add CLIMutable Type

**Comment by Zaid-Ajaj** [comment]
Very interesting, ofcourse you can become a contributer! 
I need to take a close look to this PR in the next day or two. I noticed is that you sent the PR from your `master` branch, it is recommended that you make a fresh branch for each feature you add. Anoher thing I observed is that my Travis CI wasn't triggered on this PR so I need to test locally first (I will clone your repo on master). The README.md is no big deal, we can fix manually :smile:

**Comment by Zaid-Ajaj** [comment]
@humhei I added you as a collaborator :)

**Comment by humhei** [comment]
I am sorry . My 126 mailbox didn't receive your mail
I have just switched 126 mail to outlook.
please add me again :smile:

**Comment by humhei** [comment]
https://github.com/humhei/Suave.OData.LiteDB
i have I added you as a collaborator too :smile:

**Comment by Zaid-Ajaj** [comment]
@humhei No problem, invitation sent again to you :smile: I should be able to publish a new nuget package with your updates in the next couple of days

**Comment by humhei** [comment]
Thank you,Now I have recivied the email

发件人: Zaid Ajaj<mailto:notifications@github.com>
发送时间: 2017年11月26日 21:23
收件人: Zaid-Ajaj/LiteDB.FSharp<mailto:LiteDB.FSharp@noreply.github.com>
抄送: humhei<mailto:humhei@outlook.com>; Mention<mailto:mention@noreply.github.com>
主题: Re: [Zaid-Ajaj/LiteDB.FSharp] Add DbRef Feature And Test (#2)


@humhei<https://github.com/humhei> No problem, invitation sent again to you 😄 I should be able to publish a new nuget package with your updates in the next couple of days

**Pull request merged**